ethexe/cli/src/tests.rs

```diff
@@ -0,0 +1,309 @@
+        )?;
+
+        let sender_address = sender_public_key.to_address();
+        let validators = vec![validator_public_key.to_address()];
+        let ethereum = Ethereum::deploy(&rpc, validators, signer.clone(), sender_address).await?;
+        let blob_reader = Arc::new(MockBlobReader::default());
+
+        let router_address = ethereum.router().address();
+
+        let router_query = RouterQuery::new(&rpc, router_address).await?;
+        let genesis_block_hash = router_query.genesis_block_hash().await?;
+
+        let query = Query::new(
+            Box::new(db.clone()),
+            &rpc,
+            router_address,
+            genesis_block_hash,
+            blob_reader.clone(),
+            10000,
+        )
+        .await?;
+
+        let processor = Processor::new(db.clone())?;
+
+        let sequencer = Sequencer::new(
+            &ethexe_sequencer::Config {
+                ethereum_rpc: rpc.clone(),
+                sign_tx_public: sequencer_public_key,
+                router_address,
+            },
+            signer.clone(),
+        )
+        .await?;
+
+        let validator = Validator::new(
+            &ethexe_validator::Config {
+                pub_key: validator_public_key,
+                router_address,
+            },
+            signer.clone(),
+        );
+
+        let observer = Observer::new(&rpc, router_address, blob_reader.clone())
+            .await
+            .expect("failed to create observer");
+
+        let rpc = ethexe_rpc::RpcService::new(9090, db.clone());
+
+        let service = Service::new_from_parts(
+            db.clone(),
+            network,
+            observer.clone(),
+            query,
+            processor,
+            signer,
+            Some(sequencer),
+            Some(validator),
+            None,
+            rpc,
+        );
+
+        let env = TestEnv {
+            _db: db,
+            blob_reader,
+            observer,
+            ethereum,
+            _router_query: router_query,
+            service: Some(service),
+        };
+
+        Ok(env)
+    }
+
+    pub fn new_listener(&self) -> Listener {
+        Listener::new(self.observer.clone())
+    }
+
+    pub async fn upload_code(&self, code: &[u8]) -> Result<(H256, CodeId)> {
+        let (tx_hash, code_id) = self
+            .ethereum
+            .router()
+            .upload_code_with_sidecar(code)
+            .await?;
+        self.blob_reader
+            .add_blob_transaction(tx_hash, code.to_vec())
+            .await;
+        Ok((tx_hash, code_id))
+    }
+}
+
+#[tokio::test(flavor = "multi_thread")]
+#[ntest::timeout(60_000)]
+async fn ping() {
+    let _ = env_logger::try_init();
+
+    let anvil = Anvil::new().try_spawn().unwrap();
+
+    let mut env = TestEnv::new(anvil.ws_endpoint()).await.unwrap();
+    let mut listener = env.new_listener();
+
+    let service = env.service.take().unwrap();
+    let _ = tokio::task::spawn(service.run());
```

Comment:
would be nice to join this gracefully

Reply:
Add abort presently. May be will make some better solution in future.